### PR TITLE
Add a mechanism to limit the features built during development to reduce build time.

### PR DESCRIPTION
**Pull request by ranjeshj**
Adding a props file to exclude features that you don't want to build during development. If a developer is working on a single feature like say ItemsRepeater, we don't need to spend time building all the rest of the features. This should drastically cut the build time and help speed up inner loop. The way this is done is by adding a new solution file (MuxControlsInnerLoop.sln) and a props file which gets used only when loaded through the new solution. The new solution also removes a bunch of projects not commonly used like the xaml islands projects, cx projects and adhoc project. 

This currently shows the usage to get feedback. I have not coded up for all features yet. 

The lab builds will still go through muxcontrols.sln which should not be affected in any way by this change.

### CODE REVIEW

File: dev/dll/Microsoft.UI.Xaml.vcxproj [hunk 1]
```diff
@@ -93,13 +94,13 @@
     <Import Project="..\Scroller\Scroller.vcxitems" Label="Shared" />
     <Import Project="..\ScrollViewer\ScrollViewer.vcxitems" Label="Shared" Condition="$(BuildLeanMuxForTheStoreApp) != 'true'" />
     <Import Project="..\Lights\Lights.vcxitems" Label="Shared" />
-    <Import Project="..\Repeater\Repeater.vcxitems" Label="Shared" />
+    <Import Project="..\Repeater\Repeater.vcxitems" Label="Shared" Condition="$(RepeaterExcluded) != 'true'"/>
```
Comment (jevansaks) [inline dev/dll/Microsoft.UI.Xaml.vcxproj]: My thinking here was that the .props file would have a list of directory names to be included and then we could use MSBuild wildcard expansion to find all *.vcxitems under that set of directories and then pass that list to `<Import/>` (according to https://github.com/microsoft/msbuild/pull/2014) Import tag should allow an item list to be passed to it.

Does that make sense? I'm not entirely sure if it would work, but that seems a bit more maintainable than a list of conditions that needs to be managed in the vcxproj too.

File: FeatureAreaFilterForInnerLoop.props [hunk 2]
```diff
@@ -0,0 +1,51 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT License. See LICENSE in the project root for license information. -->
+<Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (kmahone) [inline FeatureAreaFilterForInnerLoop.props]: We should probably come at this from the other direction. By default InnerLoop should build a very bare-bones MUX with almost everything excluded. An IC can then add the specific feature they wish to work on to the build.

File: FeatureAreaFilterForInnerLoop.props [hunk 3]
```diff
@@ -0,0 +1,51 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (kmahone) [inline FeatureAreaFilterForInnerLoop.props]: We should update the Create New Control script to update this file also.

File: dev/Scroller/APITests/ScrollerLayoutTests.cs [hunk 4]
```diff
@@ -80,8 +80,8 @@ public void BasicMargin()
                 SnapPointsMode.Ignore,
                 hookViewChanged: true,
                 isAnimationsEnabledOverride: null,
-                c_defaultUIScrollerContentWidth + 2 * c_Margin - c_defaultUIScrollerWidth,
-                c_defaultUIScrollerContentHeight + 2 * c_Margin - c_defaultUIScrollerHeight);
+                expectedFinalHorizontalOffset: c_defaultUIScrollerContentWidth + 2 * c_Margin - c_defaultUIScrollerWidth,
```
Comment (kmahone) [inline dev/Scroller/APITests/ScrollerLayoutTests.cs]: Are these changes related to the main PR?

File: FeatureAreaFilterForInnerLoop.props (hunk 2 above)
Comment (ranjeshj) [reply FeatureAreaFilterForInnerLoop.props]: That can trip up someone who does not know about this feature no ?

File: dev/dll/Microsoft.UI.Xaml.vcxproj [hunk 5]
```diff
@@ -93,13 +94,13 @@
...
     <Import Project="..\SwipeControl\SwipeControl.vcxitems" Label="Shared" Condition="$(BuildLeanMuxForTheStoreApp) != 'true'" />
     <Import Project="..\TwoPaneView\TwoPaneView.vcxitems" Label="Shared" Condition="$(BuildLeanMuxForTheStoreApp) != 'true'" />
     <Import Project="..\TestHooks\TestHooks.vcxitems" Label="Shared" />
     <Import Project="..\PullToRefresh\PTRTracing\PTRTracing.vcxitems" Label="Shared" Condition="$(BuildLeanMuxForTheStoreApp) != 'true'" />
     <Import Project="..\SplitButton\SplitButton.vcxitems" Label="Shared" Condition="$(BuildLeanMuxForTheStoreApp) != 'true'" />
-    <Import Project="..\LayoutPanel\LayoutPanel.vcxitems" Label="Shared" />
+    <Import Project="..\LayoutPanel\LayoutPanel.vcxitems" Label="Shared" Condition="$(LayoutPanelExcluded) != 'true'"/>
```
Comment (kmahone) [inline dev/dll/Microsoft.UI.Xaml.vcxproj]: We should probably delete all references to BuildLeanMuxForTheStoreApp since this is a more general version of that. If we ever needed to create a similar "lean mux", it would probably take the form of a custom version of FeatureAreaFilterForInnerLoop.props.

File: dev/dll/Microsoft.UI.Xaml.vcxproj [hunk 6]
```diff
@@ -1,6 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <!-- Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT License. See LICENSE in the project root for license information. -->
 <Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$(MSBuildProjectDirectory)\..\..\FeatureAreaFilterForInnerLoop.props" />
```
Comment (kmahone) [inline dev/dll/Microsoft.UI.Xaml.vcxproj]: What controls this .props file only getting loaded in the InnerLoop.sln? Add a comment, because it is not obvious that this does not always get imported.

**Comment by kmahone** [comment]
> #if !BUILD_LEAN_MUX_FOR_THE_STORE_APP

We should update this to use reflection to determine what test pages are available.

---
Refers to: test/MUXControlsTestApp/TestInventory.cs:37 in db9dc02. [](commit_id = db9dc02839ada355d913574a8e10391c41402440, deletion_comment = False)

**Comment by kmahone** [comment]
>   <Import Project="..\..\dev\ScrollViewer\InteractionTests\ScrollViewer_InteractionTests.projitems" Label="Shared" Condition="$(BuildLeanMuxForTheStoreApp) != 'true' And $(BuildingWithBuildExe) != 'true'" />

You can also delete all references to BuildingWithBuildExe.

Getting rid of both that and BuildLeanMuxForTheStoreApp should simply many of the Conditions a lot.

---
Refers to: test/MUXControls.Test/MUXControls.Test.Shared.targets:10 in db9dc02. [](commit_id = db9dc02839ada355d913574a8e10391c41402440, deletion_comment = False)

File: dev/dll/Microsoft.UI.Xaml.vcxproj [hunk 7]
```diff
@@ -111,8 +112,8 @@
     <Import Project="..\Interactions\SliderInteraction\SliderInteraction.vcxitems" Label="Shared" Condition="$(UseInternalSDK) == 'true'" />
     <Import Project="..\MenuFlyout\MenuFlyout.vcxitems" Label="Shared" />
     <Import Project="..\TeachingTip\TeachingTip.vcxitems" Label="Shared" Condition="$(BuildLeanMuxForTheStoreApp) != 'true'" />
-    <Import Project="..\RadioMenuFlyoutItem\RadioMenuFlyoutItem.vcxitems" Label="Shared" Condition="$(BuildLeanMuxForTheStoreApp) != 'true'" />
-    <Import Project="..\AnimatedVisualPlayer\AnimatedVisualPlayer.vcxitems" Label="Shared" Condition="$(BuildLeanMuxForTheStoreApp) != 'true' and $(UseInsiderSDK) == 'true'" />
+    <Import Project="..\RadioMenuFlyoutItem\RadioMenuFlyoutItem.vcxitems" Label="Shared" Condition="$(BuildLeanMuxForTheStoreApp) != 'true'" /> 
+    <Import Project="..\AnimatedVisualPlayer\AnimatedVisualPlayer.vcxitems" Label="Shared" Condition="$(BuildLeanMuxForTheStoreApp) != 'true' and $(UseInsiderSDK) == 'true' and $(AnimatedVisualPlayerExcluded) != 'true'" />
```
Comment (kmahone) [inline dev/dll/Microsoft.UI.Xaml.vcxproj]: We should fix the use of UseInsiderSDK. Right now we are always building with the 19H1 SDK and UseInsiderSDK is set to true. We should remove the UseInsiderSDK check from AnimatedVisualPlayer, since it should always build. 
At some point UseInsiderSDK will likely mean a 19H2 sdk, so we should keep it around. 

Sorry for all the "cleanup" suggestions, I just want to avoid the "Conditions" getting out of hand.

File: dev/dll/Microsoft.UI.Xaml.vcxproj (hunk 1 above)
Comment (ranjeshj) [reply dev/dll/Microsoft.UI.Xaml.vcxproj]: Expansion inside Import does not seem to work. I started a thread with the person who added the feature. If i provide the list inline, it works, but not through an ItemGroup.

File: FeatureAreaFilterForInnerLoop.props (hunk 2 above)
Comment (kmahone) [reply FeatureAreaFilterForInnerLoop.props]: It seems like the most common case is that I am working on a particular feature (e.g. ColorPicker), so I would only want to build that feature. It would be easier to enable ColorPicker than disable everything except ColorPicker. That seems like the most common use case for this, so we should make that easy.

File: dev/dll/Microsoft.UI.Xaml.vcxproj (hunk 1 above)
Comment (ranjeshj) [reply dev/dll/Microsoft.UI.Xaml.vcxproj]: Unfortunately I learnt that this is not possible in msbuild after a long email thread. Will try to simplify a little more.

File: dev/Scroller/APITests/ScrollerLayoutTests.cs (hunk 4 above)
Comment (ranjeshj) [reply dev/Scroller/APITests/ScrollerLayoutTests.cs]: These errors started showing up in the build for some reason and i had to fix them.

File: test/MUXControlsTestApp/TestInventory.cs [hunk 8]
```diff
@@ -34,58 +34,165 @@ static TestInventory()
         {
             Tests = new List<TestDeclaration>()
             {
-#if !BUILD_LEAN_MUX_FOR_THE_STORE_APP
+// #if FEATURE_SCROLLER_ENABLED && FEATURE_COLOR_PICKER_ENABLED && FEATURE_NAVIGATION_VIEW_ENABLED && FEATURE_RATING_CONTROL_ENABLED
```
Comment (ranjeshj) [inline test/MUXControlsTestApp/TestInventory.cs]: To be fixed

File: test/MUXControlsTestApp/MUXControlsTestApp.Shared.projitems [hunk 9]
```diff
@@ -157,9 +164,12 @@
   </ItemGroup>
   <ItemGroup>
     <Compile Include="$(MSBuildThisFileDirectory)VisualTreeTestHelper.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)\LeakTestPage.xaml.cs" Condition="$(BuildLeanMuxForTheStoreApp) != 'true'">
+    
+    <!-- Need to make this more modular. It is kind of mixing different features together in one page -->
+    <!--Compile Include="$(MSBuildThisFileDirectory)\LeakTestPage.xaml.cs" Condition="$(BuildLeanMuxForTheStoreApp) != 'true'">
```
Comment (ranjeshj) [inline test/MUXControlsTestApp/MUXControlsTestApp.Shared.projitems]: To be fixed

File: test/MUXControlsTestApp/TestInventory.cs [hunk 10]
```diff
@@ -34,58 +34,165 @@ static TestInventory()
...
+//              {"Leak", typeof(LeakTestPage)},
+//#endif
+
+#if FEATURE_COLOR_PICKER_ENABLED
```
Comment (ranjeshj) [inline test/MUXControlsTestApp/TestInventory.cs]: I think i'm going to change these names to match the ones in the native (like COLORPICKER_INCLUDED) or change the other ones to match this. Let me know if you prefer one over the other.

File: test/MUXControlsTestApp/MUXControlsTestApp.Shared.projitems [hunk 11]
```diff
@@ -133,7 +135,12 @@
     <NoWarn>$(NoWarn);0105;8305</NoWarn>
   </PropertyGroup>
   <PropertyGroup Condition="$(BuildLeanMuxForTheStoreApp) == 'true'">
-    <DefineConstants>$(DefineConstants);BUILD_LEAN_MUX_FOR_THE_STORE_APP</DefineConstants>
+    <DefineConstants>$(DefineConstants);BUILD_LEAN_MUX_FOR_THE_STORE_APP</DefineConstants> 
+  </PropertyGroup>
+  <PropertyGroup>
+    <DefineConstants Condition="$(AnimatedVisualPlayerExcluded) != 'true'">$(DefineConstants);ANIMATED_VISUAL_PLAYER_INCLUDED</DefineConstants>
+    <DefineConstants Condition="$(RepeaterExcluded) != 'true'">$(DefineConstants);REPEATER_INCLUDED</DefineConstants>
+    <DefineConstants Condition="$(LayoutPanelExcluded) != 'true'">$(DefineConstants);LAYOUT_PANEL_INCLUDED</DefineConstants>
```
Comment (ranjeshj) [inline test/MUXControlsTestApp/MUXControlsTestApp.Shared.projitems]: remove these

File: FeatureAreaFilterForInnerLoop.props (hunk 2 above)
Comment (ranjeshj) [reply FeatureAreaFilterForInnerLoop.props]: I added a list of entries, we can just enable the ones that you care about.

File: dev/dll/Microsoft.UI.Xaml.vcxproj (hunk 6 above)
Comment (ranjeshj) [reply dev/dll/Microsoft.UI.Xaml.vcxproj]: updated it to load always.

File: dev/dll/Microsoft.UI.Xaml.vcxproj [hunk 12]
```diff
@@ -70,62 +71,65 @@
...
+    <Import Project="..\Repeater\Repeater.vcxitems" Label="Shared" Condition="$(FeatureRepeaterEnabled) == 'true'" />
+    <Import Project="..\SwipeControl\SwipeControl.vcxitems" Label="Shared" Condition="$(FeatureSwipeControlEnabled) == 'true'" />
+    <Import Project="..\TwoPaneView\TwoPaneView.vcxitems" Label="Shared" Condition="$(FeatureTwoPaneViewEnabled) == 'true'" />
+    <Import Project="..\PullToRefresh\PTRTracing\PTRTracing.vcxitems" Label="Shared" Condition="$(FeaturePullToRefreshEnabled) == 'true'" />
+    <Import Project="..\SplitButton\SplitButton.vcxitems" Label="Shared" Condition="$(FeatureSplitButtonEnabled) == 'true'" />
+    <Import Project="..\LayoutPanel\LayoutPanel.vcxitems" Label="Shared" Condition="$(FeatureLayoutPanelEnabled) == 'true'" />
+    <Import Project="..\DropDownButton\DropDownButton.vcxitems" Label="Shared" Condition="$(FeatureDropDownButtonEnabled) == 'true'" />
+    <Import Project="..\RadioButtons\RadioButtons.vcxitems" Label="Shared" Condition="$(FeatureRadioButtonsEnabled) == 'true'" />
     <!-- IconSource is implemented in Windows.UI.Xaml.dll in the OS repo, so we don't want to include it on that side. -->
```
Comment (kmahone) [inline dev/dll/Microsoft.UI.Xaml.vcxproj]: Delete this obsolete comment.

File: dev/dll/Microsoft.UI.Xaml.vcxproj [hunk 13]
```diff
@@ -70,62 +71,65 @@
...
-    <Import Project="..\IconSource\IconSource.vcxitems" Label="Shared" />
+    <Import Project="..\IconSource\IconSource.vcxitems" Label="Shared" Condition="$(FeatureIconSourceEnabled) == 'true'" />
     <!-- These two depend on the type InteractionBase, which is behind the Velocity feature Feature_Xaml2018 in the OS repo.
```
Comment (kmahone) [inline dev/dll/Microsoft.UI.Xaml.vcxproj]: Delete this obsolete comment.

File: dev/dll/CommandingHelpers.cpp [hunk 14]
```diff
@@ -110,16 +109,16 @@ winrt::IInspectable CommandingHelpers::WUXIconSourceToMUXIconSourceConverter::Co
     winrt::throw_hresult(E_NOTIMPL);
 }
 
-void CommandingHelpers::BindToLabelPropertyIfUnset(
+void CommandingHelpers::BindToIconSourcePropertyIfUnset(
```
Comment (kmahone) [inline dev/dll/CommandingHelpers.cpp]: What's the purpose of these changes?

File: dev/NavigationView/NavigationView.vcxitems [hunk 15]
```diff
@@ -9,6 +9,7 @@
   <ItemDefinitionGroup>
     <ClCompile>
       <AdditionalIncludeDirectories>%(AdditionalIncludeDirectories);$(MSBuildThisFileDirectory)</AdditionalIncludeDirectories>
+      <PreprocessorDefinitions>%(PreprocessorDefinitions);NAVIGATIONVIEW_INCLUDED</PreprocessorDefinitions>
```
Comment (kmahone) [inline dev/NavigationView/NavigationView.vcxitems]: Why can these definitions not go in the common FeatureAreas.props?

**Comment by kmahone** [comment]
Since this changes gets rid of BuildLeanMuxForTheStoreApp, please do a search across the codebase and remove all references to it (search for BuildLeanMuxForTheStoreApp, BUILDLEANMUXFORTHESTOREAPP and BUILD_LEAN_MUX_FOR_THE_STORE_APP).

File: test/MUXControlsTestApp/TestInventory.cs [hunk 16]
```diff
@@ -34,58 +34,165 @@ static TestInventory()
...
                 {"ColorPicker", typeof(ColorPickerPage), "ColorPicker.png"},
-                {"Leak", typeof(LeakTestPage)},
+#endif
+
+#if FEATURE_PERSON_PICTURE_ENABLED
```
Comment (kmahone) [inline test/MUXControlsTestApp/TestInventory.cs]: Consider doing this with reflection instead of a bunch of ifdefs. You could create a new attribute [WinUITestPage] (or whatever) then apply that to all the top-level test pages. Then do reflection across the types to find them.

If you don't want to do it, please file a GitHub issue. You can even assign it to me. :)

File: dev/dll/CommandingHelpers.cpp (hunk 14 above)
Comment (kmahone) [reply dev/dll/CommandingHelpers.cpp]: Ah, ok. The two functions just switched place in the file.

---
In reply to: [302729511](https://github.com/microsoft/microsoft-ui-xaml/pull/956#discussion_r302729511) [](ancestors = 302729511)

File: dev/AnimatedVisualPlayer/TestUI/AnimatedVisualPlayerPage.xaml.cs [hunk 17]
```diff
@@ -38,6 +38,7 @@ public FallbackGrid()
     /// <summary>
     /// An empty page that can be used on its own or navigated to within a Frame.
     /// </summary>
+    [AddToTestInventory(Name = "AnimatedVisualPlayer", Icon = "Animations.png")]
```
Comment (kmahone) [inline dev/AnimatedVisualPlayer/TestUI/AnimatedVisualPlayerPage.xaml.cs]: Conventionally attributes are declaratively named. So this could be something like [TopLevelTestPage] or something like that.

**Comment by kmahone** [comment]
> [CmdLetBinding()]

Update this script to add the new attribute to the test page.

---
Refers to: tools/GenerateNewControlProject.ps1:1 in c6b16d7. [](commit_id = c6b16d78862177538dd8596c3b58670febca0521, deletion_comment = False)

File: test/MUXControlsTestApp/TestInventory.cs [hunk 18]
```diff
@@ -3,6 +3,8 @@
 
 using System;
 using System.Collections.Generic;
+using System.Reflection;
+using Windows.UI.Notifications;
```
Comment (kmahone) [inline test/MUXControlsTestApp/TestInventory.cs]: What is this namespace used for?

File: test/MUXControlsTestApp/TestInventory.cs (hunk 10 above)
Comment (ranjeshj) [reply test/MUXControlsTestApp/TestInventory.cs]: removed most of then, I think we can avoid these now that we are using reflection.

---
In reply to: [302677541](https://github.com/microsoft/microsoft-ui-xaml/pull/956#discussion_r302677541) [](ancestors = 302677541)

File: dev/NavigationView/NavigationView.vcxitems (hunk 15 above)
Comment (ranjeshj) [reply dev/NavigationView/NavigationView.vcxitems]: I like that these are packaged along with the projects and not outside of them. This can be done for the mux dll since all of them build into one. For the tests I had to create them in the props file but I'm getting rid of them in the props file altogether and using reflection instead.

---
In reply to: [302730563](https://github.com/microsoft/microsoft-ui-xaml/pull/956#discussion_r302730563) [](ancestors = 302730563)

File: test/MUXControlsTestApp/TestInventory.cs (hunk 18 above)
Comment (ranjeshj) [reply test/MUXControlsTestApp/TestInventory.cs]: removed
